### ISSUE TITLE
Add forward_backward_custom method to TrainingClient

### ISSUE THREAD

**Issue by tang-t21** [nex-agi/weaver#40]
## Summary

Add a `forward_backward_custom` method to `TrainingClient` that allows users to define arbitrary differentiable loss functions on logprobs, using the surrogate loss trick (inspired by Tinker).

## Background

Currently loss functions are identified by string (`"cross_entropy"`, `"importance_sampling"`, etc.). This feature enables user-defined custom loss functions that run locally on the client while gradients are computed on the GPU cluster.

### Architecture

```
Client                          Server (no changes)         Trainer
  |                                                           |
  |-- forward_backward("forward_logprob") ------------------>|
  |<-------------- logprobs per sample ----------------------|
  |                                                           |
  | [run user's loss_fn locally]                              |
  | [loss.backward() -> grad weights]                         |
  |                                                           |
  |-- forward_backward("surrogate", surrogate_weights) ----->|
  |<-------------- gradients accumulated --------------------|
```

Trade-off: ~1.5x FLOPs, up to 3x wall time vs built-in loss functions (extra forward pass required).

## API

```python
def forward_backward_custom(
    self,
    data: Sequence[Datum],
    loss_fn: Callable[[Sequence[Datum], List[torch.Tensor]], Tuple[torch.Tensor, Dict[str, Any]]],
) -> Dict[str, Any]:
```

User's `loss_fn` receives `(data, logprobs_list)` and returns `(loss_tensor, metrics_dict)`.

### Modify: `weaver/training_client.py`

Implementation steps:
1. Call `forward_backward(data, "forward_logprob")` to get logprobs
2. Parse logprobs into `torch.Tensor` with `requires_grad=True` (reuse pattern from `pig_latin.py:69-75`)
3. Call user's `loss_fn(data, logprob_tensors)` -> `(loss, metrics)`
4. `loss.backward()` to get `d(loss)/d(logprobs)`
5. Build surrogate `Datum` objects with `surrogate_weights = logprob.grad`
6. Call `forward_backward(surrogate_data, "surrogate")` for backward pass
7. Return `{"loss": loss.detach(), "metrics": metrics}`

### Error handling
- Empty/mismatched `loss_fn_outputs` -> ValueError
- User's `loss_fn` failure -> RuntimeError
- Missing `.grad` after backward -> RuntimeError
- Non-scalar loss -> validate `loss.dim() == 0`

## Testing

- Mock `forward_backward` to return canned logprobs
- Verify loss_fn called with correct tensors, gradients extracted, surrogate Datum built correctly
- Integration test: simple custom loss + optim_step, verify training progresses

## Dependencies

Requires `"surrogate"` loss type in weaver-trainer: china-qijizhifeng/weaver-trainer#99
